reapers/endpoints.go

```diff
@@ -5,430 +5,265 @@ import (
-const maxRetries = 10
-const retryWait = time.Second * 2
-
-// Magic number: I took a guess. DO NOT consider this number authoritative
-const reapLimit uint = 20
-
 type EndpointReaper struct {
-	nomad       *nomad_api.Client
-	consul      *consul_api.Client
-	cilium      *client.Client
-	ctx         context.Context
-	cidr        *net.IPNet
-	numToReap   uint
-	excludeTags []string
+	cilium           EndpointUpdater
+	nomadAllocations AllocationInfo
+	nomadEventStream EventStreamer
 }
 
 // NewEndpointReaper creates a new EndpointReaper. This will run an initial reconciliation before
 // returning the reaper
-func NewEndpointReaper(ctx context.Context, nomad_client *nomad_api.Client, consul_client *consul_api.Client, cidr *net.IPNet, excludeTags []string) (*EndpointReaper, error) {
-	// TODO: Improve this once we figure out how we are setting up cilium everywhere
-	cilium, err := client.NewDefaultClient()
-	if err != nil {
-		return nil, fmt.Errorf("error when connecting to cilium agent: %s", err)
+func NewEndpointReaper(ciliumClient EndpointUpdater, nomadAllocations AllocationInfo, nomadEventStream EventStreamer) (*EndpointReaper, error) {
+	reaper := EndpointReaper{
+		cilium:           ciliumClient,
+		nomadAllocations: nomadAllocations,
+		nomadEventStream: nomadEventStream,
 	}
-	reaper := EndpointReaper{nomad: nomad_client, consul: consul_client, cidr: cidr, cilium: cilium, ctx: ctx, excludeTags: excludeTags}
+
 	// Do the initial reconciliation loop
 	if err := reaper.reconcile(); err != nil {
 		return nil, fmt.Errorf("unable to perform initial reconciliation: %s", err)
 	}
+
 	return &reaper, nil
 }
 
 // Run the reaper until the context given in the contructor is cancelled. This function is non
 // blocking and will only return errors if something occurs during startup
 // return a channel to notify of consul client failures
-func (e *EndpointReaper) Run() (<-chan bool, error) {
+func (e *EndpointReaper) Run(ctx context.Context) (<-chan bool, error) {
+
 	// NOTE: Specifying uint max so that it starts from the next available index. If there is a
 	// better way to start from latest index, we can change this
-	eventChan, err := e.nomad.EventStream().Stream(e.ctx, map[nomad_api.Topic][]string{nomad_api.TopicJob: {}}, math.MaxInt64, &nomad_api.QueryOptions{Namespace: "*"})
+	eventChan, err := e.nomadEventStream.Stream(
+		ctx,
+		map[nomad_api.Topic][]string{
+			nomad_api.TopicJob: {},
+		},
+		math.MaxInt64,
+		&nomad_api.QueryOptions{
+			Namespace: "*",
+		},
+	)
 	if err != nil {
 		return nil, fmt.Errorf("error when starting node event stream: %s", err)
 	}
+
 	failChan := make(chan bool, 1)
 
 	go func() {
-		tick := time.NewTicker(time.Hour)
-		defer tick.Stop()
-
 		for {
 			select {
-			case <-e.ctx.Done():
-				zap.S().Info("Context cancelled, shutting down endpoint reaper")
+			case <-ctx.Done():
+				zap.L().Info("Context cancelled, shutting down endpoint reaper")
 				return
-			case <-tick.C:
```

Comment:
Is there a reason you got rid of the reconciliation tick? I worry about some sort of missed allocation event that could make it so we are out of sync

Reply:
Mostly simplicity and we've not seen dropped events so far. However I don't think having it is a problem so I'll restore it.